### PR TITLE
Enable storing and loading backdoors

### PR DESCRIPTION
**Pull request by VRehnberg**
A variant of storing and loading backdoors as discussed in https://github.com/ejnnr/cupbearer/issues/4

I didn't find a nice way of handling this for all transforms so only doing it for backdoors now. Will probably need a larger make-over sometime in the future.

This is probably the PR that deviates the most with respect to coding style so far, so feel free to be especially harsh in the code review.

### CODE REVIEW

**Comment by VRehnberg** [comment]
Also I couldn't decide whether I should reproduce most of the previous tests with wanet just to see if loading worked in both train_detector and eval_detector. Or if it would be better to just switch out use of corner and wanet from how it is now.

File: src/cupbearer/scripts/eval_detector.py [hunk 1]
```diff
@@ -15,6 +16,11 @@ def main(cfg: Config):
         save_dir=cfg.dir.path,
     )
 
+    if not cfg.no_load and hasattr(cfg.task, "backdoor"):
```
Comment (ejnnr) [inline src/cupbearer/scripts/eval_detector.py:19]: I'm a bit worried this could break silently (e.g. on a refactor that changes the `backdoor` attribute). I know I have a few similar checks elsewhere, just trying to figure out if there's a better way here.

Is there any reason the approach I mentioned in #4 wouldn't work? I.e. run this `load` method inside `BackdoorDetection._get_anomalous_test_data()` and use the path that's passed to `BackdoorDetection` anyway. Then `no_load` could be an attribute on `BackdoorDetection` instead, and we wouldn't need the `hasattr` check

File: src/cupbearer/scripts/train_detector.py [hunk 2]
```diff
@@ -15,6 +15,10 @@ def main(cfg: Config):
         rng=jax.random.PRNGKey(cfg.seed),
     )
 
+    if not cfg.no_load and hasattr(cfg.task, "backdoor"):
```
Comment (ejnnr) [inline src/cupbearer/scripts/train_detector.py:18]: Similar to comment above: we might be able to move this into the `BackdoorDetection` task

**Review by ejnnr — COMMENTED** [review]
Thanks, overall looks good! I left a few comments, the main one is whether we can do the loading inside `BackdoorDetection` to avoid the `hasattr` checks (which feel a bit brittle). But I might be missing a reason that wouldn't work.

For tests: I think swapping wanet and corner is better because otherwise the test will just take longer and longer to run, and that isn't going to scale. But I think the pipeline test is due some rewrites anyway (at the very least using fixtures that are global to the module, so we can split it up into several functions without sacrificing speed). So as long as you've checked manually that it runs, I wouldn't worry about the test too much for now.

Let me know once I should take another look!

File: src/cupbearer/data/backdoors.py [hunk 3]
```diff
@@ -106,6 +107,27 @@ def __post_init__(self):
         p_transform = self.p_backdoor + self.p_noise
         assert 0 <= p_transform <= 1, "Probability must be between 0 and 1"
 
+    @staticmethod
+    def _get_savefile_fullpath(basepath):
+        return os.path.join(basepath, "wanet_backdoor.npy")
+
+    def store(self, basepath):
+        super().store(basepath)
+        # TODO If img size is known the transform can be significantly sped up
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/data/backdoors.py:116]: It would require knowing the size of the image before starting training, but I figure I could just make it a lazy property and then store and load the full field.

[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/data/backdoors.py:116]: Only calculating the full warping field once is the key thing for speed-up. So I guess storing and loading the 4x4 warping matrices is fine if we want to keep all doors open.

File: src/cupbearer/data/backdoors.py [hunk 4]
```diff
@@ -106,6 +107,27 @@ def __post_init__(self):
...
+        # by pre-computing and saving the full flow field instead.
+        np.save(self._get_savefile_fullpath(basepath), self.control_grid)
+
+    def load(self, basepath):
+        super().load(basepath)
+        self.control_grid = np.load(self._get_savefile_fullpath(basepath))
+
+        # TODO might be okay to just update control_grid_width with a warning
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/data/backdoors.py:124]: Yes, that sounds reasonable and should be easy enough.

File: src/cupbearer/scripts/eval_detector.py (hunk 1 above)
Comment (VRehnberg) [reply src/cupbearer/scripts/eval_detector.py:19]: I'll take a look. I remember looking at that function previously, don't know if there was a reason why I didn't put it there.

File: src/cupbearer/scripts/train_classifier.py [hunk 5]
```diff
@@ -115,6 +115,10 @@ def main(cfg: Config):
     )
     if cfg.dir.path:
         trainer.save_model()
+        try:
```
[the remark this answers is not in the document]
Comment (VRehnberg) [reply src/cupbearer/scripts/train_classifier.py:118]: I had some issue with that I remember. But, I'll try again.

**Comment by VRehnberg** [comment]
I'll probably not have much time to work on this the coming few weeks, so if you find yourself waiting on this, feel free to wrap it up yourself.

File: src/cupbearer/data/backdoors.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/cupbearer/data/backdoors.py:116]: Yeah, I don't think it matters much which one we store on disk (I guess storing the full field has the advantage that it also deals with the `control_grid_width` thread below, but that's easy to fix anyway).

Agree that computing the field once we get the first image and then caching is the best way to do this

**Comment by ejnnr** [comment]
Finished this up and did some testing, I'm fairly sure it works as intended now. Abstractions detector gets better than random performance now (but only ~0.8 AUROC), Mahalanobis not really. This was only a quick MNIST run though, planning to do actual experiments with this soon.
